### DISCUSSION TITLE
Is RavenDB Embedded as slow as it seems??

### DISCUSSION THREAD

**Discussion by 3threeFX** [ravendb/ravendb#13707] [General]
In response to the question posed in the title, I hope this is not true, because I really like this platform.
According to the recent issue I posted #13666 I realized that some basic operations are done noticeably slower than SQLite.
For more clarity, I have visualized this comparison in multi scenarios as described below:   

### Scenario 1: Outside the coding environment 

Raven Studio v5.3.1 (Embedded) and DB Browser (SQLite) V3.12.2: the result shows in a simple instruction (for projection and distinct data) SQLite is 20x ~ 30x faster! (SQLite: 70 ~78 ms, RavenDB Embedded: 1,875 ~ 2,743 ms) In other words, the Query run time has been improved in SQLite by 90% ~ 95% compared to RavenDB.

Database file size: 
Raven Studio v5.3.1 (Embedded): compact + without compression level => 250 MB 
Raven Studio v5.3.1 (Embedded): compact + Document Compression =>128MB 
SQLite : VACUUM => 24MB

![2022-02-24_15-30-23](https://user-images.githubusercontent.com/63404512/155520102-5549a6e2-2498-433f-8ad0-101a71cecfc4.png)

![SQLitedb](https://user-images.githubusercontent.com/63404512/155517529-58d938d3-ec63-4a99-a6e0-8a097efd3207.gif)

![ravendb_studio_ez](https://user-images.githubusercontent.com/63404512/155517951-41b36613-d8fc-4517-a2af-f7821e8dd72d.gif)

but if we just do a simple projection without a distinct method, The conditions for both are relatively competitive (RavenDB Embedded: 21ms, SQLite: 25 ms).
Although this issue seems illogical in the first place and this might sound a bit stupid, considering the simplification of the comparison scenarios and the similarity of the conditions applied to both platforms, maybe it will be worth considering and discussing.

**Comment by ayende** [comment]
Next time, please send the details, instead of animated GIFs. It made it very hard to understand what is going on.
Detailed answer to your questions:

* RavenDB should be of comparable performance to SQLite for most operations. However, for certain actions, each engine may chose to do that differently, which can have performance differences.
* Size on disk - RavenDB will allocate disk size by default in 64MB increments. You can Compact the database if you want to free the data it reserved, but this is done to optimize performance and avoid disk fragmentation.


As for your questions, a query such as `select distinct field` forces RavenDB to scan through all the documents, project the value and then return just the unique values. A similar query, like so `from Symbols group by symbol_type select symbol_type`, on the other hand, will allow RavenDB to do the computation once and return the results far faster.

**Comment by 3threeFX** [comment]
Dear @ayende first of all thanks for your quick response 💐🙏 secondly contrary to your opinion, I think visual representation helps to achieve common literature and increase mutual understanding But to respect your request, I will avoid this.
I have uploaded the exported .csv file of the database to complete the documentation of these comparisons [[here ](https://file.io/a2H6Zkv9pRFN)] ...

In response to the second part of your answer, as I said before with compact + compression => Raven.voron size on disk is about 130 MB ...

I tested your suggestion query :
`from 'Symbols' group by symbol_type select symbol_type`

![2022-02-24_16-57-21](https://user-images.githubusercontent.com/63404512/155542983-6fc9fce2-96fa-474e-aa73-655868e6aa7f.png)


The duration time => 2,822 ms **for the first time**. (after that because auto-indexing it decreased noticeably < 10 ms)
There is still a significant difference with SQLite performance (**at first run** without considering Auto Index Map-Reduce)

**Comment by 3threeFX** [comment]
### Scenario 2: inside C# ToList(); Tragedy!

Considering the similarity of the conditions applied to both platforms these snippet codes were executed and observed that irrationally Serialize/Deserialize documents make ToList() method very slow (I think related to [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/)) 

**Simple Projection:**

ToList() in Raven Studio v5.3.1 (Embedded) ~ 4 s  [[debug](https://filebin.net/bissyp26url4agar/RDBtolist.gif)] 
ToDynamicList() Linq.Dynamic.Core and [EF Core+SQLite](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore.Sqlite/) : 400~500ms [[debug](https://filebin.net/bissyp26url4agar/sqlitetolist.gif)] 

![2022-02-24_17-36-14](https://user-images.githubusercontent.com/63404512/155542777-ca26d9c5-c8d9-4b11-bca5-5a08bebbcaf8.png)

However, applying operators (In such dimensions of data ~250K samples) such as Distinct(), Where(), OrderBy(), again ToList() etc. to the output list takes less than 50ms with LINQ!

**Comment by ayende** [comment]
### Scenario 1

The first time you issue an index, we have to do significant amount of work around that. We don't actually care for that scenario, since we are optimizing for ongoing costs, not a single run. Over time, even without creating indexes, RavenDB will learn and create the most efficient system for your needs.

Also note that you are comparing us in a tabular data format, where SQLite has a built-in advantage. You are doing something that it is quite well suited for and you are seeing good results, unsurprisingly. 




### Scenario 2

You are comparing apples and oranges here, I'm afraid.
I'm not sure if you noticed, but in your SQLite query, you are running the same `distinct` approach as before. However, what you are doing with RavenDB is quite different.

You are asking RavenDB to give you _all_ the `Symbols` that you have, and you are doing that using a query method.
What this means is that:

RavenDB has to return ~250,000 (!) results. 
What you are actually doing is to send a request to get all those results as a single batch, so we build that on the RavenDB side, send that to the client, where it reads all of that, materialize all 250K result and hand that back to you.

That is a significantly more work done than I believe you were expecting.

**Comment by 3threeFX** [comment]
### Scenario 2

It seems that a misunderstanding has happened to you as I said before I appreciate that following the [debug] links in the last post.
so what we know about this situation:
### Assumptions and findings :
1) It was a simple projection in SQLite and Ravendb.
2) I intentionally split the code into several sections to better determine how much time each section spends.
3) The output in both compared codes is => SIZE(total data) ~ 250K row x 1 column

Snippet code without any distinct or ... in SQLite:

```
 List<string> resultList = new List<string>();
            using (var db = new MetadataModel())
            {
                var result = db.Symbols.AsNoTracking().AsQueryable();
                var result_2 = result.Select(columnName).AsParallel();
                var result_3 = result_2.ToDynamicList<string>().AsParallel();
            }
```
4) As you can see in the beneath image SQLite has returned almost near 250K (**indistinct results** !!) in 350ms (elapsed time ->only for the last line that List<string> result_3 created) it's not a wired thing at all! but it's wired when the exact line (I emphasize exact line) for creating a list in Ravendb long much more (near 4s) ... 

`var result_3 = result_2.ToList();`

![new](https://user-images.githubusercontent.com/63404512/155736313-11410606-2909-4356-9316-9730083fd2e7.png)

5) It is worth mentioning that our simple operation runs faster in Raven Studio but wrapping these outputs with .ToList() method has a different story. 

![vs](https://user-images.githubusercontent.com/63404512/155747370-e525b088-4cac-4fec-a87a-883c7cf1603c.png)

6) I believe Due to its storage nature, this time overhead is imposed on the system ([Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/)). 

7) Actually, I don't need this scenario (it was done only for comparison and benchmarking and nothing else) but because of different approaches in distinct and ..., I try to compare just projection operation in one field (symbol_type) and try to get a string list to compare **both of them in a fair method**.

### Conclusion [ Apple :? Orange => Banana😂] :
It can be deduced that Ravendb do select operation faster than SQLite but wrapping results to a list is tragedic! **Ravendb returned 250K x 1 data in 4,000 ms and SQLite has returned identical 250K x 1 in 350 ms** 
So dear @ayende Is this conclusion acceptable based on the assumptions made or not? Or do you still believe that there is something wrong with this comparison and that it is unfairly designed? 
Don't hesitate, I'm glad to hear your opinion.💐🙏

**Reply by ayende** [reply]
What you are actually doing in both cases is very different.

RavenDB isn't actually running in-process, like SQLite, you are running as a child process. You are also doing something that we explicitly recommend against (reading large amount of results via `ToList`). That requires us to do buffering in the server and the client process.

If you want to get all of them, you will need to use the Streaming API, which is dedicated for that scenario.

See: https://ravendb.net/docs/article-page/5.3/csharp/client-api/session/querying/how-to-stream-query-results#session-querying-how-to-stream-query-results

There is also the serialization overhead, yes.

**Comment by kamranayub** [comment]
This should probably move to a discussion since this seems more about specific code being executed. A benchmark should be for a real-world use case and I'm not sure I see one here? There is overheard for JSON deserialization, yes, but it all depends on the specific use case to decide on ways to model it differently or more efficiently. I think this is what Ayende is referring to as apples vs. oranges. There are a lot of things you have to "unlearn" when coming from SQL-based solutions (speaking from experience migrating to RavenDB from EF a long time ago).

**Comment by 3threeFX** [comment]
> This should probably move to a discussion since this seems more about specific code being executed. A benchmark should be for a real-world use case and I'm not sure I see one here? There is overheard for JSON deserialization, yes, but it all depends on the specific use case to decide on ways to model it differently or more efficiently. I think this is what Ayende is referring to as apples vs. oranges. There are a lot of things you have to "unlearn" when coming from SQL-based solutions (speaking from experience migrating to RavenDB from EF a long time ago).

Dear @kamranayub , thanks for your response. Almost all of the things you mentioned are also acceptable to me (Especially specific code/data used here) but as can be inferred from @ayende 's writings There was a misunderstanding in the comparison process for him, I just emphasized that both platforms have been tested with the same condition and the same amount of indistinct data under load and I think apple and oranges referenced to this misunderstanding.

By the way, thank you for participating in this discussion and also for your great advice.

**Reply by kamranayub** [reply]
To compare closer performance characteristics, I think you would need to store that dataset in SQLite using JSON structure and use some JSON functions. So you'd still need to deserialize the results back.

https://www.sqlite.org/json1.html

The entire data format is different for this specific case and it's unlikely you'd actually load 250k documents in one call, especially in a user-facing use case. For batch processing, you'd use Data Subscriptions or streaming more likely, not a query like this. By default queries limit the result set to 1024 documents for safety since it's not expected you'd do this under normal circumstances.

But again, it really depends on what you're attempting to do in the exact context you require.

**Reply by 3threeFX** [reply]
> To compare closer performance characteristics, I think you would need to store that dataset in SQLite using JSON structure and use some JSON functions. So you'd still need to deserialize the results back.
> 
> https://www.sqlite.org/json1.html
> 
> The entire data format is different for this specific case and it's unlikely you'd actually load 250k documents in one call, especially in a user-facing use case. For batch processing, you'd use Data Subscriptions or streaming more likely, not a query like this. By default queries limit the result set to 1024 documents for safety since it's not expected you'd do this under normal circumstances.
> 
> But again, it really depends on what you're attempting to do in the exact context you require.

As I mentioned before it's related to the last issue #13666, You might think that I have high expectations of Ravendb, but in that topic, obviously, it was still noticeably slow in tests. 
maybe in a regular app, it's odd but in data mining or Machine Learning, this size of data is not strange at all...
as I told you before serializing/deserializing documents reduced performance and certainly json+sqlite => It will have the same bottleneck! but @ayende Attributes that to other causes however I accepted that part of @ayende 's talk, the comparative inherent advantages of one platform over another must be acknowledged, but the huge difference in performance has taken me by surprise Even in situations where the main engine has better power.

For the records, I read many users complain of slow writing speed (The period of previous years, not recent years) in Ravendb but my experience tells me that it's much faster than SQLite with EF Core (SQLite : 25 s , Ravendb:7 s for 250k) however I used bulk Insert method, not regular mode.
